### ISSUE TITLE
[add2app] --dart-define has no effects when building module

### ISSUE THREAD

**Issue by zmtzawqlp** [flutter/flutter#76672]
I am trying to add flutter into existing android/ios appliaction.
and build aar/framework as following command

```
flutter build aar --dart-define=flag=abc
flutter build ios-framework --dart-define=flag=abc
```

and i found i can't get value from following code.
```
String.fromEnvironment('flag', defaultValue: null);

```

```
[✓] Flutter (Channel stable, 1.22.5, on macOS 11.0.1 20B29 darwin-x64, locale zh-Hans-CN)
    • Flutter version 1.22.5 at /Users/zmtzawqlp/Documents/flutter/stable
    • Framework revision 7891006299 (3 months ago), 2020-12-10 11:54:40 -0800
    • Engine revision ae90085a84
    • Dart version 2.10.4
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/zmtzawqlp/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/zmtzawqlp/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin installed
    • Dart plugin version 201.9245
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.53.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.19.0
```

</details>

**Comment by zmtzawqlp** [comment]
on android, i change flutter_tools to support --dart-define 
https://github.com/flutter/flutter/pull/71829/files

**Comment by zmtzawqlp** [comment]
i have download demo from https://github.com/flutter/samples/tree/master/add_to_app/fullscreen
and change config to use aar/framework

android

1.go into flutter_module project, change code as following. run 'flutter build aar --dart-define=flag=abc' 
``` dart
class FullScreenView extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('${String.fromEnvironment('flag', defaultValue: null)}'),
      ),
      body: const Contents(showExit: true),
    );
  }
}
```
2.go into android_fullscreen 
change app/build.gradle as following
``` 
apply plugin: 'com.android.application'

apply plugin: 'kotlin-android'

apply plugin: 'kotlin-android-extensions'

android {
    compileSdkVersion 28
    defaultConfig {
        applicationId "dev.flutter.example.androidfullscreen"
        minSdkVersion 19
        targetSdkVersion 28
        versionCode 1
        versionName "1.0"
        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
        multiDexEnabled true
    }
    buildTypes {
        release {
            minifyEnabled false
        }
    }
    compileOptions {
        sourceCompatibility 1.8
        targetCompatibility 1.8
    }
}

String storageUrl = System.env.FLUTTER_STORAGE_BASE_URL ?: "https://storage.googleapis.com"
repositories {
    maven {
        url '/Users/zmt/Documents/flutter samples master add_to_app-fullscreen/flutter_module/build/host/outputs/repo'
    }
    maven {
       url '$storageUrl/download.flutter.io'
    }
}

dependencies {
    implementation 'androidx.multidex:multidex:2.0.1'
 //   implementation project(':flutter')
    debugImplementation 'dev.flutter.example.flutter_module:flutter_debug:1.0'
//    profileImplementation 'dev.flutter.example.flutter_module:flutter_profile:1.0'
    releaseImplementation 'dev.flutter.example.flutter_module:flutter_release:1.0'

    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
    implementation 'androidx.appcompat:appcompat:1.1.0'
    implementation 'androidx.core:core-ktx:1.1.0'
    implementation 'androidx.constraintlayout:constraintlayout:1.1.3'
    testImplementation 'junit:junit:4.12'
    androidTestImplementation project(':espresso')
    androidTestImplementation "com.google.truth:truth:0.42"
    androidTestImplementation 'androidx.test:runner:1.1.1'
    androidTestImplementation 'androidx.test.espresso:espresso-core:3.1.0'
    api 'androidx.test:core:1.2.0'
}
```

change settings.gradle as following
```
include ':app'
include ':espresso'
//rootProject.name='AndroidFullScreen'
//setBinding(new Binding([gradle: this]))
//evaluate(new File(
//        settingsDir.parentFile,
//        'flutter_module/.android/include_flutter.groovy'
//))
//include ':flutter_module'
//project(':flutter_module').projectDir = new File('../flutter_module')

```
3.run android_fullscreen, and go to flutter page, you can see 'null' in AppTitle

**Comment by zmtzawqlp** [comment]
it doesn't work in master branch too.
```
Flutter 1.27.0-5.0.pre.96 • channel master • https://github.com/flutter/flutter.git
Framework • revision dac5e8af54 (2 days ago) • 2021-02-22 13:44:19 +0000
Engine • revision b793775d2a
```
Tools • Dart 2.13.0 (build 2.13.0-52.0.dev)

**Comment by TahaTesser** [comment]
Hi @zmtzawqlp 
Thanks for filing the issue, I can reproduce using `flutter/samples`, I suspect this is not probably not supported yet

<details>
<summary>flutter doctor -v</summary>

```bash
[✓] Flutter (Channel stable, 1.22.6, on macOS 11.2.1 20D74 darwin-x64, locale en-GB)
    • Flutter version 1.22.6 at /Users/tahatesser/Code/flutter_stable
    • Framework revision 9b2d32b605 (5 weeks ago), 2021-01-22 14:36:39 -0800
    • Engine revision 2f0af37152
    • Dart version 2.10.5
[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, set ANDROID_SDK_ROOT to that location.
      You may also want to add it to your PATH environment variable.
[!] Xcode - develop for iOS and macOS (Xcode 12.4)
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    ✗ CocoaPods not installed.
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To install:
        sudo gem install cocoapods
[!] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
[✓] VS Code (version 1.53.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.19.0
[!] Connected device
    ! No devices available
! Doctor found issues in 4 categories.
```
```bash
[✓] Flutter (Channel master, 1.27.0-5.0.pre.134, on macOS 11.2.1 20D74 darwin-x64, locale en-GB)
    • Flutter version 1.27.0-5.0.pre.134 at /Users/tahatesser/Code/flutter_master
    • Framework revision 354bd53f62 (4 hours ago), 2021-02-24 02:36:05 -0500
    • Engine revision 9061823775
    • Dart version 2.13.0 (build 2.13.0-68.0.dev)
[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, please use
      `flutter config --android-sdk` to update to that location.
[!] Xcode - develop for iOS and macOS
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    ✗ CocoaPods not installed.
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To install see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.
[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome
[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
[✓] VS Code (version 1.53.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.19.0
[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 11.2.1 20D74 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 88.0.4324.192
! Doctor found issues in 2 categories.
```

</details>

**Comment by zmtzawqlp** [comment]
It's very strange that we support the command but it's not working actually. It's very important for who's app is add2app.

**Comment by AlexV525** [comment]
https://github.com/flutter/flutter/pull/71829 includes tests that ensure it works, IMO this must be a misunderstanding in the build procedure or the test is fake. /cc @jmagman

**Comment by jmagman** [comment]
Hmm I thought this was done with #52044.

**Comment by jmagman** [comment]
Oh, #71829 isn't on stable, it's available on the beta channel.  Can you confirm this works on the beta channel?

**Comment by AlexV525** [comment]
> Can you confirm this works on the beta channel?

Well the master channel can repo this.

**Comment by jmagman** [comment]
@AlexV525 and @zmtzawqlp you both say it doesn't work on the master channel, you mean both `flutter build aar` and `flutter build ios-framework`?  They are hooked up in different ways.

**Comment by AlexV525** [comment]
According to our online chat records, at lease aar doesn't work as expected.

**Comment by zmtzawqlp** [comment]
| Flutter version | Platform | Test Case      | Result   |
| --------------- | -------- | -------------- | -------- |
| 1.22.6          | Android  | MyApp          | Bad      |
| 1.22.6          | Ios      | MyApp          | Bad      |
| 1.22.6          | Android  | Flutter/Sample | Bad      |
| 1.22.6          | Ios      | Flutter/Sample | Not test |
| 1.27            | Android  | Flutter/Sample | Bad      |
| 1.27            | Ios      | Flutter/Sample | Not test |

**Comment by zmtzawqlp** [comment]
ios

1.go into flutter_module project, change code as following. run 'flutter build  ios-framework --dart-define=flag=abc' 
``` dart
class FullScreenView extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('${String.fromEnvironment('flag', defaultValue: null)}'),
      ),
      body: const Contents(showExit: true),
    );
  }
}
```
2.change profile, comment flutter config
```
# Uncomment the next line to define a global platform for your project
# platform :ios, '9.0'

#flutter_application_path = '../flutter_module'
#load File.join(flutter_application_path, '.ios', 'Flutter', 'podhelper.rb')

target 'IOSFullScreen' do
  # Comment the next line if you don't want to use dynamic frameworks
  use_frameworks!

  # Pods for IOSFullScreen
 # install_all_flutter_pods(flutter_application_path)

  target 'IOSFullScreenTests' do
    inherit! :search_paths
    # Pods for testing
  end

  target 'IOSFullScreenUITests' do
    inherit! :search_paths
    # Pods for testing
  end

end
```
3.go into  flutter samples master add_to_app-fullscreen/flutter_module/build/ios/framework/debug
 drag framework into Build Phases > Link Binary With Libraries

4.In the target’s build settings, add $(PROJECT_DIR)/Flutter/Release/ to the Framework Search Paths (FRAMEWORK_SEARCH_PATHS).

5. General=> Frameworks,Libraies,and Embedded Content
Embed & Sign framework

6.pod install, run ios_fullscreen, click button, Appbar title is show null.

![image](https://user-images.githubusercontent.com/16477333/109109565-7d782380-7770-11eb-8058-3d8d4572b4fa.png)

**Comment by DennisLit** [comment]
Flutter 2.0 listed --dart-define for aar in the changelog as a new supported feature so we built some functionality depending upon --dart-define for our modules and then only now realizing they don't actually work as listed above, please do fix.

**Comment by TahaTesser** [comment]
This seems to be fixed for iOS 



I tried to reproduce the issue on the `stable`, tested on add2app, the passed param never shows up, however, this seems to be fixed in later channels 

### Steps to reproduce 
1. `git clone https://github.com/flutter/samples.git`
2. Modify `amples/tree/master/add_to_app/prebuilt_module/flutter_module`add following lines in the code in the column
```dart
             Text(const String.fromEnvironment('MY_VAR')),
             Text(const String.fromEnvironment('MY_OTHER_VAR')),
```
3. Build framework
`flutter build ios-framework --dart-define=MY_VAR=MY_VALUE --dart-define=MY_OTHER_VAR=MY_OTHER_VALUE  --no-universal --output=../ios_using_prebuilt_module/Flutter/`

4. Launch the native iOS app `ios_using_prebuilt_module`

### Preview
| `Channel stable, 2.0.5` | `beta`, `dev`, `master`  |
| --------------- | --------------- |
<img src="https://user-images.githubusercontent.com/48603081/116705509-470b8e80-a9d5-11eb-9804-c822e26c4cd2.PNG" height="450" /> | <img src="https://user-images.githubusercontent.com/48603081/116705516-47a42500-a9d5-11eb-88d2-c871d89661d7.PNG" height="450" /> |


<details> 
<summary>flutter doctor -v</summary> 
	
```console
[✓] Flutter (Channel stable, 2.0.5, on macOS 11.3 20E232 darwin-x64, locale en-GB)
    • Flutter version 2.0.5 at /Users/tahatesser/Code/flutter_stable
    • Framework revision adc687823a (2 weeks ago), 2021-04-16 09:40:20 -0700
    • Engine revision b09f014e96
    • Dart version 2.12.3

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Volumes/Extreme/SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Volumes/Extreme/SDK
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.5, Build version 12E262
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.55.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (3 available)
    • Taha’s iPhone (mobile) • 00008020-001059882212002E • ios            • iOS 14.5
    • macOS (desktop)        • macos                     • darwin-x64     • macOS 11.3 20E232 darwin-x64
    • Chrome (web)           • chrome                    • web-javascript • Google Chrome 90.0.4430.93

• No issues found!

```
		
</details>

Can you please upgrade to the latest  `beta` and try again
`flutter channel beta`
`flutter upgrade --force`
`flutter doctor -v`

**Comment by jmagman** [comment]
Bisected, this was fixed in #75027 which is currently on the beta channel.
Thanks for the easy reproduction steps, @TahaTesser.

Closing.
